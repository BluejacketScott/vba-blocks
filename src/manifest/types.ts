export type Version = string;

export interface DependencyDetails {
  name: string;
  version?: string;
}

export interface RegistryDependency extends DependencyDetails {
  registry: string;
  version: string;
}

export interface PathDependency extends DependencyDetails {
  path: string;
}

export interface GitDependency extends DependencyDetails {
  git: string;
  tag?: string;
  branch?: string;
  rev?: string;
}

export type Dependency = RegistryDependency | PathDependency | GitDependency;

export interface ReferenceDetails {
  dependency?: string;
}

export interface Reference {
  name: string;
  version: string;
  guid: string;
  major: number;
  minor: number;
  optional?: boolean;
  details?: ReferenceDetails;
}

export interface Source {
  name: string;
  path: string;
  binary?: string;
  optional?: boolean;
  original?: string;
}

export type TargetType = 'xlsx' | 'xlsm' | 'xlam';

export interface Target {
  name: string;
  type: TargetType;
  path: string;
  filename: string;
  blank?: boolean;
}

export interface Snapshot {
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Metadata {
  authors: string[];
  publish: boolean;
  [name: string]: any;

  __temp_defaults?: string[];
}

export type ManifestType = 'package' | 'project';

export interface Manifest extends Snapshot {
  type: ManifestType;
  metadata: Metadata;
  src: Source[];
  references: Reference[];
  target?: Target;
  dir: string;
}
