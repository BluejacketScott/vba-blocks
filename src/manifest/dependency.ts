import { manifestOk } from '../errors';
import has from '../utils/has';
import { isString } from '../utils/is';
import { extname, join, trailing } from '../utils/path';
import { Version } from './version';

/*
  # Dependency

  Three types of dependencies: registry, path, and git

  - registry: version | { version, registry? }

    Registry (default = "vba-blocks"):
    index + packages are loading in config and used for resolve and fetching

  - path: path
  - git: { git, branch?, tag?, rev? }

    Where git is path to git remote and revision is pulled from
    master, branch, tag, or rev
*/

export interface DependencyDetails {
  name: string;
  version?: Version;
}

export interface RegistryDependency extends DependencyDetails {
  registry: string;
  version: Version;
}

export interface PathDependency extends DependencyDetails {
  path: string;
}

export interface GitDependency extends DependencyDetails {
  git: string;
  tag?: string;
  branch?: string;
  rev?: string;
}

export type Dependency = RegistryDependency | PathDependency | GitDependency;

const EXAMPLE = `Example vba-block.toml:

  [dependencies]
  a = "^1.0.0"
  b = { version = "^0.1.0" }
  c = { path = "packages/c" }
  d = { git = "https://github.com/author/d" }
  e = { git = "https://github.com/author/e", branch = "next" }
  f = { git = "https://github.com/author/f", tag = "v1.0.0" }
  g = { git = "https://github.com/author/g", rev = "a1b2c3d4" }

  [dependencies.h]
  version = "^2.0.0"`;

export function parseDependencies(value: any, dir: string): Dependency[] {
  return Object.entries(value).map(([name, value]) => parseDependency(name, value, dir));
}

export function parseDependency(name: string, value: Version | any, dir: string): Dependency {
  if (isString(value)) value = { version: value };

  let {
    registry = 'vba-blocks',
    version,
    path,
    git,
    tag,
    branch = 'master',
    rev
  }: {
    registry?: string;
    version?: string;
    path?: string;
    git?: string;
    tag?: string;
    branch?: string;
    rev?: string;
  } = value;

  manifestOk(
    version || path || git,
    `Invalid dependency "${name}", no version, path, or git specified. \n\n${EXAMPLE}`
  );

  if (version) {
    return { name, registry, version };
  } else if (path) {
    return { name, path: trailing(join(dir, path)) };
  } else {
    git = ensureGitUrl(git!);

    if (rev) return { name, git, rev };
    else if (tag) return { name, git, tag };
    else return { name, git, branch };
  }
}

export function isRegistryDependency(dependency: Dependency): dependency is RegistryDependency {
  return has(dependency, 'registry');
}

export function isPathDependency(dependency: Dependency): dependency is PathDependency {
  return has(dependency, 'path');
}

export function isGitDependency(dependency: Dependency): dependency is GitDependency {
  return has(dependency, 'git');
}

export function formatDependencies(dependencies: Dependency[]): object {
  const value: { [name: string]: any } = {};
  dependencies.forEach(dependency => {
    if (isRegistryDependency(dependency)) {
      const { name, registry, version } = dependency;
      value[name] = registry !== 'vba-blocks' ? { version, registry } : version;
    } else if (isPathDependency(dependency)) {
      const { name, path } = dependency;
      value[name] = path;
    } else {
      const { name, git, tag, branch, rev } = dependency;
      value[name] = { name, git };
      if (tag) value[name].tag = tag;
      if (branch) value[name].branch = branch;
      if (rev) value[name].rev = rev;
    }
  });

  return value;
}

function ensureGitUrl(value: string): string {
  if (extname(value!) === '.git') return value;

  return `${value}.git`;
}
