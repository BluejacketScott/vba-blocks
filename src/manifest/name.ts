import dedent from '@timhall/dedent/macro';
import { CliError, ErrorCode } from '../errors';

/*
  # Project/package name

  Name requirements:
  - Must not be empty
  - Allowed characters: A-Z a-z 0-9 - _ . /
  - Must start with an alphanumeric
  - A single / is allowed for scoped packages
  - A maximum length of 255 for the scope and 255 for the package name

  Note:
  - Capitals are allowed, but compared and stored as lowercase
*/

const ONLY_VALID_CHARACTERS = /^[A-Za-z0-9-_./]*$/;
const STARTS_WITH_ALPHANUMERIC = /^[A-Za-z0-9]/;
const MAX_LENGTH = 255;

export interface ManifestName {
  scope: string | undefined;
  name: string;
  parts: string[];
}

export function parseName(name: string): ManifestName {
  assertValidName(name);

  // Manifest name is case-insensitve,
  // but parsed name isn't for comparison, uniqueness, and file naming
  name = name.toLowerCase();

  let scope: string | undefined;
  let pkg: string;
  if (name.includes('/')) {
    [scope, pkg] = name.split('/');
  } else {
    pkg = name;
  }

  return { scope, name: pkg, parts: scope ? [scope, pkg] : [pkg] };
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
}

export function validateName(name: string): ValidationResult {
  const errors = [];

  if (!name || typeof name !== 'string' || !name.length) {
    errors.push('name must not be empty');
  }
  if (!ONLY_VALID_CHARACTERS.test(name)) {
    errors.push('name must only include the following characters: A-Z a-z 0-9 - _ . /');
  }
  if (!STARTS_WITH_ALPHANUMERIC.test(name)) {
    errors.push('name must only start with the following characters: A-Z a-z 0-9');
  }
  if (countMatches(name, /\//g) > 1) {
    errors.push('name must only include a single / for scoped packages');
  }

  let scope: string | undefined;
  let pkg: string;
  if (name.includes('/')) {
    [scope, pkg] = name.split('/');
  } else {
    pkg = name;
  }

  if (scope && scope.length > MAX_LENGTH) {
    errors.push(`scope must be less than ${MAX_LENGTH} characters`);
  }
  if (pkg.length > MAX_LENGTH) {
    errors.push(`name must be less than ${MAX_LENGTH} characters`);
  }

  return { valid: errors.length === 0, errors };
}

export function assertValidName(name: string) {
  const { valid, errors } = validateName(name);
  if (valid) return;

  const error_messages = errors.map(error => `- ${error}`).join('\n');

  throw new CliError(
    ErrorCode.ManifestNameInvalid,
    dedent`
      Invalid project/package name "${name}"

      Errors:
      ${error_messages}
    `
  );
}

function countMatches(value: string, search: RegExp): number {
  return (value.match(search) || []).length;
}
