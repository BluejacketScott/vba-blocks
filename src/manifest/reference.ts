import { manifestOk } from '../errors';

/*
  # Reference

  To add reference, guid, major version, and minor version are required

  reference: { guid, version }

  Where version is "MAJOR.MINOR"
*/

export interface ReferenceDetails {
  dependency?: string;
}

export interface Reference {
  name: string;
  guid: string;
  major: number;
  minor: number;
}

const VERSION_REGEX = /^(\d+)\.(\d+)$/;
const GUID_REGEX = /\{.{8}-.{4}-.{4}-.{4}-.{12}\}/;

const isVersion = (value: string) => VERSION_REGEX.test(value);
const isGuid = (value: string) => GUID_REGEX.test(value);

const toInt = (value: string) => parseInt(value, 10);
const getMajorMinor = (version: string) => {
  const [major, minor] = version.split('.', 2).map(toInt);

  return { major, minor };
};

const EXAMPLE = `Example vba-block.toml:

  [references.Scripting]
  version = "1.0"
  guid = "{420B2830-E718-11CF-893D-00A0C9054228}"`;

export function parseReferences(value: any): Reference[] {
  return Object.entries(value).map(([name, value]) => parseReference(name, value));
}

export function parseReference(name: string, value: any): Reference {
  const { version, guid } = value;

  manifestOk(
    isVersion(version),
    `Reference "${name}" has an invalid version "${version}". \n\n${EXAMPLE}.`
  );
  manifestOk(isGuid(guid), `Reference "${name}" has an invalid guid "${guid}". \n\n${EXAMPLE}'`);

  const { major, minor } = getMajorMinor(version);

  return { name, guid, major, minor };
}
