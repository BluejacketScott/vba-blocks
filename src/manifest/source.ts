import { manifestOk } from '../errors';
import { isString } from '../utils/is';
import { join } from '../utils/path';

/*
  # Source

  A source file to be imported

  source: path | { path }
*/

export interface Source {
  name: string;
  path: string;
  binary?: string;
}

const EXAMPLE = `Example vba-block.toml:

  [src]
  A = "src/a.bas"
  B = { path = "src/b.cls" }`;

export function parseSrc(value: any, dir: string): Source[] {
  return Object.entries(value).map(([name, value]) => parseSource(name, value, dir));
}

export function parseSource(name: string, value: string | any, dir: string): Source {
  if (isString(value)) value = { path: value };
  const { path: relativePath, binary } = value;

  manifestOk(relativePath, `src "${name}" is missing path. \n\n${EXAMPLE}`);
  const path = join(dir, relativePath);

  const source: Source = { name, path };
  if (binary) source.binary = join(dir, binary);

  return source;
}
