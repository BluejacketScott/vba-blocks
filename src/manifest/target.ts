import { manifestOk } from '../errors';
import has from '../utils/has';
import { isString } from '../utils/is';
import { join, relative, sanitize } from '../utils/path';

/*
  # Target

  target: type | { type, name?, path? }

  By default "target" is used for path
*/

export type TargetType = 'xlsx' | 'xlsm' | 'xlam';

export interface Target {
  name: string;
  type: TargetType;
  path: string;
  filename: string;
}

const target_types = ['xlsx', 'xlsm', 'xlam'];

const EXAMPLE = `Example vba-block.toml:

  [project]
  target = "xlsm"

Example vba-block.toml with alternative path:

  [project]
  target = { type = "xlsm", path = "target/xlsm" }`;

export function parseTarget(value: any, pkgName: string, dir: string): Target {
  if (isString(value)) value = { type: value };
  if (!has(value, 'name')) value = { name: pkgName, ...value };
  const { type, name, path: relativePath = 'target' } = value;

  manifestOk(isString(type), `Target is missing "type". \n\n${EXAMPLE}.`);
  manifestOk(
    isSupportedTargetType(type),
    `Unsupported target type "${type}". Only "xlsx", "xlsm", and "xlam" are supported currently.`
  );

  const path = join(dir, relativePath);
  const filename = `${sanitize(name)}.${type}`;

  return { name, type, path, filename };
}

export function isSupportedTargetType(type: string): type is TargetType {
  return isString(type) && target_types.includes(type);
}

export function formatTarget(target: Target, default_name: string, dir: string): string | object {
  let { name, type: target_type, path } = target;
  path = relative(dir, path);

  let value: string | { type: string; name?: string; path?: string };
  if (name !== default_name || path !== 'target') {
    value = { type: target_type };
    if (name !== default_name) value.name = name;
    if (path !== 'target') value.path = path;
  } else {
    value = target_type;
  }

  return value;
}
