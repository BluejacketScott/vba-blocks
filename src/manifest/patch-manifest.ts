import dedent from 'dedent';
import { relative } from '../utils/path';
import { Manifest, Source, Dependency, Target } from './';
import { isRegistryDependency, isPathDependency } from './dependency';
import {
  patchApplyChanges,
  patchAddSrc,
  patchRemoveSrc,
  patchAddDependency,
  patchRemoveDependency
} from '../messages';
import env from '../env';

export function applyChanges(changes: string[]) {
  if (!changes.length) return;

  env.reporter.log(`${patchApplyChanges()}\n\n${changes.join('\n\n')}`);
}

export function addSource(manifest: Manifest, source: Source): string {
  const relative_path = relative(manifest.dir, source.path);
  const relative_binary_path = source.binary && relative(manifest.dir, source.binary);

  const details = relative_binary_path
    ? `{ path = "${relative_path}", binary = "${relative_binary_path}" }`
    : `"${relative_path}"`;

  return dedent`
    ${patchAddSrc()}
    ${source.name} = ${details}`;
}

export function removeSource(_: Manifest, name: string): string {
  return patchRemoveSrc(name);
}

export function addDependency(manifest: Manifest, dependency: Dependency): string {
  let details;
  if (isRegistryDependency(dependency)) {
    const { version, registry } = dependency;
    details =
      registry !== 'vba-blocks'
        ? `{ version = "${version}", registry = "${registry}" }`
        : `"${version}"`;
  } else if (isPathDependency(dependency)) {
    const relative_path = relative(manifest.dir, dependency.path);
    details = `{ path = "${relative_path}" }`;
  } else {
    const { git, tag, branch, rev } = dependency;
    if (rev) {
      details = `{ git = "${git}", rev = "${rev}" }`;
    } else if (tag) {
      details = `{ git = "${git}", tag = "${tag}" }`;
    } else if (branch !== 'master') {
      details = `{ git = "${git}", branch = "${branch}" }`;
    } else {
      details = `{ git = "${git}" }`;
    }
  }

  return dedent`
    ${patchAddDependency()}
    ${dependency.name} = ${details}`;
}

export function removeDependency(_: Manifest, name: string): string {
  return patchRemoveDependency(name);
}

export function addTarget(manifest: Manifest, target: Target): string {
  const relative_path = relative(manifest.dir, target.path);

  let details;
  if (target.name !== manifest.name && relative_path !== 'target') {
    details = target.type;
  } else {
    details = `{ type = "${target.type}"`;
    if (target.name !== manifest.name) details += `, name = "${target.name}"`;
    if (relative_path !== 'target') details += `, path = "${relative_path}"`;
    details += ' }';
  }

  return dedent`
    Add the following to the [${manifest.type}] section:
    target = ${details}`;
}
