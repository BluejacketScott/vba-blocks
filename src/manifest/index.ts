import { join, normalize, relative } from '../utils/path';
import { parse as parseToml, convert as convertToToml } from '../utils/toml';
import { pathExists, readFile, writeFile } from '../utils/fs';
import { manifestNotFound } from '../errors';
import { Version } from './version';
import { Source, parseSrc } from './source';
// TODO #features
// import { Feature, parseFeatures } from './feature';
import { Dependency, parseDependencies } from './dependency';
import { Reference, parseReferences } from './reference';
import { Target, parseTargets } from './target';
import { manifestOk, manifestInvalid } from '../errors';

export {
  Version,
  Source,
  // TODO #features
  // Feature,
  Dependency,
  Reference,
  Target
};

/**
 * @example
 * ```toml
 * [package]
 * name = "package-name"
 * version = "1.0.0-rc.1"
 * authors = ["Tim Hall <tim.hall.engr@gmail.com> (https://github.com/timhall)"]
 *
 * [src]
 * A = "src/a.bas"
 * B = { path = "src/b.cls" }
 * C = { path = "src/c.frm", optional = true }
 *
 * [dependencies]
 * dictionary = "v1.4.1"
 * with-properties = { version = "1.0.0" }
 * from-path = { path = "packages/from-path" }
 * from-git-master = { git = "https://github.com/VBA-tools/VBA-Web.git" }
 * from-git-branch = { git = "https://github.com/VBA-tools/VBA-Web.git", branch = "beta" }
 * from-git-tag = { git = "https://github.com/VBA-tools/VBA-Web.git", tag = "v1.0.0" }
 * from-git-rev = { git = "https://github.com/VBA-tools/VBA-Web.git", rev = "a1b2c3d4" }
 *
 * # [references.Scripting]
 * # version = "1.0"
 * # guid = "{420B2830-E718-11CF-893D-00A0C9054228}"
 *
 * [targets]
 * xlsm = "targets/xlsm"
 *
 * [targets.xlam]
 * name = "custom-name"
 * path = "targets/xlam"
 * ```
 */

export interface Snapshot {
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Metadata {
  authors: string[];
  publish: boolean;
  [name: string]: any;
}

export interface Manifest extends Snapshot {
  type: ManifestType;
  metadata: Metadata;
  src: Source[];
  references: Reference[];
  targets: Target[];
  dir: string;

  // TODO #features
  // features: Feature[];
  // defaultFeatures: string[];
}

export type ManifestType = 'package' | 'project';

const EXAMPLE = `Example vba-block.toml for a package (e.g. library to be shared):

  [package]
  name = "my-package"
  version = "0.0.0"
  authors = ["..."]

Example vba-block.toml for a project (e.g. workbook, document, etc.):

  [project]
  name = "my-project"
  version = "0.0.0"
  authors = ["..."]`;

export function parseManifest(value: any, dir: string): Manifest {
  manifestOk(
    value && (value.package || value.project),
    `[package] or [project] is required, with name, version, and authors specified. \n\n${EXAMPLE}`
  );

  let type: ManifestType, name, version, authors, publish;
  if (value.project) {
    type = 'project';
    name = value.project.name;
    version = value.project.version || '0.0.0';
    authors = value.project.authors || [];
    publish = false;

    manifestOk(name, `[project] name is a required field. \n\n${EXAMPLE}`);
  } else {
    type = 'package';
    name = value.package.name;
    version = value.package.version;
    authors = value.package.authors;
    publish = value.package.publish || false;

    manifestOk(name, `[package] name is a required field. \n\n${EXAMPLE}`);
    manifestOk(
      version,
      `[package] version is a required field. \n\n${EXAMPLE}`
    );
    manifestOk(
      authors,
      `[package] authors is a required field. \n\n${EXAMPLE}`
    );
  }

  const src = parseSrc(value.src || {}, dir);
  const dependencies = parseDependencies(value.dependencies || {}, dir);
  const references = parseReferences(value.references || {});
  const targets = parseTargets(value.targets || {}, name, dir);

  // TODO #features
  // const { features, defaultFeatures } = parseFeatures(value.features || {});

  return {
    type,
    name,
    version,
    metadata: { authors, publish },
    src,
    dependencies,
    references,
    targets,
    dir

    // TODO #features
    // features,
    // defaultFeatures,
  };
}

export async function loadManifest(dir: string): Promise<Manifest> {
  const file = join(dir, 'vba-block.toml');

  if (!(await pathExists(file))) {
    throw manifestNotFound(dir);
  }

  const raw = await readFile(file);

  let parsed;
  try {
    parsed = parseToml(raw.toString());
  } catch (err) {
    const message = `Syntax Error: ${file} (${err.line}:${err.column})\n\n${
      err.message
    }`;
    throw manifestInvalid(message);
  }

  const manifest = parseManifest(parsed, normalize(dir));

  return manifest;
}

export async function writeManifest(manifest: Manifest, dir: string) {
  const { type, name, version, metadata } = manifest;
  const value: any = {
    [type]: { name, version, ...metadata }
  };

  value.src = {};
  manifest.src.forEach(source => {
    let { path, optional } = source;
    path = relative(manifest.dir, path);
    value.src[source.name] = optional ? { path, optional } : path;
  });

  value.targets = {};
  manifest.targets.forEach(target => {
    let { name, type, path } = target;
    path = relative(manifest.dir, path);
    value.targets[type] = name !== manifest.name ? { name, path } : path;
  });

  // TODO dependencies and references

  const toml = convertToToml(value);
  await writeFile(join(dir, 'vba-block.toml'), toml);
}
