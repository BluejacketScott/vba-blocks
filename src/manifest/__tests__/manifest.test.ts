import { loadManifest, parseManifest } from '../';
import { dir as FIXTURES, invalidManifest, standard } from '../../../tests/__fixtures__';
import { normalizeManifest } from '../../__helpers__/manifest';

const BASE_MANIFEST: {
  package: {
    name: string;
    version: string;
    authors: string[];
    target?: string | object;
  };
} = {
  package: { name: 'package-name', version: '1.0.0', authors: ['Tim Hall'] }
};

test('loads valid package metadata', () => {
  expect(normalizeManifest(parseManifest(BASE_MANIFEST, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid package metadata', () => {
  expect(() => parseManifest({}, FIXTURES)).toThrow();
  expect(() => parseManifest({ package: {} }, FIXTURES)).toThrow();
  expect(() => parseManifest({ package: { name: 'package-name' } }, FIXTURES)).toThrow();
  expect(() => parseManifest({ package: { version: '1.0.0' } }, FIXTURES)).toThrow();
  expect(() => parseManifest({ package: { authors: ['Tim Hall'] } }, FIXTURES)).toThrow();
});

test('loads valid sources', () => {
  const value = {
    ...BASE_MANIFEST,
    src: {
      A: 'src/a.bas',
      B: { path: 'src/b.cls' },
      C: { path: 'src/c.frm', optional: true }
    }
  };

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('loads valid [dev-src]', () => {
  const value = {
    ...BASE_MANIFEST,
    ['dev-src']: {
      A: 'src/a.bas',
      B: { path: 'src/b.cls' },
      C: { path: 'src/c.frm', optional: true }
    }
  };

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid sources', () => {
  const value = {
    ...BASE_MANIFEST,
    src: {
      missing_path: { optional: true }
    }
  };

  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('loads valid dependencies', () => {
  const value = {
    ...BASE_MANIFEST,
    dependencies: {
      a: '^1.0.0',
      b: {
        version: '^2.0.0',
        optional: true,
        features: ['a', 'b'],
        'default-features': false
      },
      c: { path: 'packages/c' },
      d: { git: 'https://github.com/VBA-tools/VBA-Web' },
      e: { git: 'https://github.com/VBA-tools/VBA-Web', branch: 'next' },
      f: { git: 'https://github.com/VBA-tools/VBA-Web', tag: 'v1.0.0' },
      g: { git: 'https://github.com/VBA-tools/VBA-Web', rev: 'a1b2c3d4' }
    }
  };

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('loads valid [dev-dependencies]', () => {
  const value = {
    ...BASE_MANIFEST,
    'dev-dependencies': {
      a: '^1.0.0',
      b: { version: '^2.0.0' },
      c: { path: 'packages/d' },
      d: { git: 'https://github.com/VBA-tools/VBA-Web' },
      e: { git: 'https://github.com/VBA-tools/VBA-Web', branch: 'next' },
      f: { git: 'https://github.com/VBA-tools/VBA-Web', tag: 'v1.0.0' },
      g: { git: 'https://github.com/VBA-tools/VBA-Web', rev: 'a1b2c3d4' }
    }
  };

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid dependencies', () => {
  const value: any = { ...BASE_MANIFEST, dependencies: { a: {} } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('loads valid references', () => {
  const value = {
    ...BASE_MANIFEST,
    references: {
      a: {
        version: '1.0',
        guid: '{420B2830-E718-11CF-893D-00A0C9054228}',
        optional: true
      }
    }
  };

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('loads valid [dev-references]', () => {
  const value = {
    ...BASE_MANIFEST,
    ['dev-references']: {
      a: {
        version: '1.0',
        guid: '{420B2830-E718-11CF-893D-00A0C9054228}',
        optional: true
      }
    }
  };

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid references', () => {
  let value: any = { ...BASE_MANIFEST, references: { a: {} } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, references: { b: { version: '1.0.0' } } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, references: { c: { version: '1.0' } } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = {
    ...BASE_MANIFEST,
    references: {
      d: { version: '1.0', guid: 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa' }
    }
  };
  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('loads valid target', () => {
  const value = {
    ...BASE_MANIFEST
  };
  value.package.target = 'xlsm';

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();

  value.package.target = { type: 'xlam', name: 'addin', path: 'targets/xlam' };

  expect(normalizeManifest(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('loads and parses manifest', async () => {
  const manifest = await loadManifest(standard);
  expect(normalizeManifest(manifest, standard)).toMatchSnapshot();
});

test('throws for invalid syntax', async () => {
  expect.assertions(1);

  try {
    await loadManifest(invalidManifest);
  } catch (err) {
    expect(err.message.replace(FIXTURES, 'fixtures')).toMatchSnapshot();
  }
});
