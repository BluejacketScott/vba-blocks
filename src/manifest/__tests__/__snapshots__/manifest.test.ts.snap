// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and parses manifest 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "json",
      "registry": "vba-blocks",
      "version": "^2",
    },
  ],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "standard",
  "references": Array [
    Object {
      "guid": "{0002E157-0000-0000-C000-000000000046}",
      "major": 5,
      "minor": 3,
      "name": "VBIDE",
    },
  ],
  "src": Array [
    Object {
      "name": "ThisWorkbook",
      "path": "src/ThisWorkbook.cls",
    },
    Object {
      "name": "Sheet1",
      "path": "src/Sheet1.cls",
    },
    Object {
      "name": "Sheet2",
      "path": "src/Sheet2.cls",
    },
    Object {
      "name": "Sheet3",
      "path": "src/Sheet3.cls",
    },
    Object {
      "name": "Validation",
      "path": "src/Validation.bas",
    },
    Object {
      "name": "Class1",
      "path": "src/Class1.cls",
    },
  ],
  "target": Object {
    "filename": "standard.xlsm",
    "name": "standard",
    "path": "targets/xlsm",
    "type": "xlsm",
  },
  "type": "project",
  "version": "0.0.0",
}
`;

exports[`loads valid [dev-dependencies] 1`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [
    Object {
      "name": "a",
      "registry": "vba-blocks",
      "version": "^1.0.0",
    },
    Object {
      "name": "b",
      "registry": "vba-blocks",
      "version": "^2.0.0",
    },
    Object {
      "name": "c",
      "path": "packages/d",
    },
    Object {
      "branch": "master",
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "d",
    },
    Object {
      "branch": "next",
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "e",
    },
    Object {
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "f",
      "tag": "v1.0.0",
    },
    Object {
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "g",
      "rev": "a1b2c3d4",
    },
  ],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "target": undefined,
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid [dev-references] 1`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [],
  "devReferences": Array [
    Object {
      "guid": "{420B2830-E718-11CF-893D-00A0C9054228}",
      "major": 1,
      "minor": 0,
      "name": "a",
    },
  ],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "target": undefined,
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid [dev-src] 1`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [
    Object {
      "name": "A",
      "path": "src/a.bas",
    },
    Object {
      "name": "B",
      "path": "src/b.cls",
    },
    Object {
      "name": "C",
      "path": "src/c.frm",
    },
  ],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "target": undefined,
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid dependencies 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "a",
      "registry": "vba-blocks",
      "version": "^1.0.0",
    },
    Object {
      "name": "b",
      "registry": "vba-blocks",
      "version": "^2.0.0",
    },
    Object {
      "name": "c",
      "path": "packages/c",
    },
    Object {
      "branch": "master",
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "d",
    },
    Object {
      "branch": "next",
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "e",
    },
    Object {
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "f",
      "tag": "v1.0.0",
    },
    Object {
      "git": "https://github.com/VBA-tools/VBA-Web.git",
      "name": "g",
      "rev": "a1b2c3d4",
    },
  ],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "target": undefined,
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid package metadata 1`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "target": undefined,
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid references 1`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [
    Object {
      "guid": "{420B2830-E718-11CF-893D-00A0C9054228}",
      "major": 1,
      "minor": 0,
      "name": "a",
    },
  ],
  "src": Array [],
  "target": undefined,
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid sources 1`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [
    Object {
      "name": "A",
      "path": "src/a.bas",
    },
    Object {
      "name": "B",
      "path": "src/b.cls",
    },
    Object {
      "name": "C",
      "path": "src/c.frm",
    },
  ],
  "target": undefined,
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid target 1`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "target": Object {
    "filename": "package-name.xlsm",
    "name": "package-name",
    "path": "target",
    "type": "xlsm",
  },
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`loads valid target 2`] = `
Object {
  "dependencies": Array [],
  "devDependencies": Array [],
  "devReferences": Array [],
  "devSrc": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "publish": undefined,
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "target": Object {
    "filename": "addin.xlam",
    "name": "addin",
    "path": "targets/xlam",
    "type": "xlam",
  },
  "type": "package",
  "version": "1.0.0",
}
`;

exports[`throws for invalid syntax 1`] = `
"vba-blocks.toml is invalid:

Syntax Error: fixtures/projects/invalid-manifest/vba-block.toml (2:6)

Error parsing TOML (2, 7):

name: Package Name
      ^
Expected \\"=\\" for key-value, found Package"
`;
